Fix interpreter ResumeAfterCatch on arm64 Unix

janvorli opened the pull request:
The `DispatchManagedException` and `DispatchRethrownManagedException` were marked as `NORETURN`. That caused a problem when the C++ compiler has optimized the invocation of such methods so that it has removed its epilog and the return address was in a completely different function. That has broken the native unwind loop in the `InterpreterCodeManager::ResumeAfterCatch`, as it restored some non-volatile registers incorrectly.

The Regressions/coreclr/GitHub_88128/test88128 was failing due to this om macOS arm64.

I have fixed it in the same way as C++ EH handles such case. The PAL_VirtualUnwind moves the PC back in case it is not a frame of a hardware exception so that when a function ends with a call followed by nothing else, it gets an instruction pointer in the function for unwinding purposes.

janvorli committed:
Fix interpreter ResumeAfterCatch on arm64 Unix

The DispatchManagedException and DispatchRethrownManagedException
were marked as NORETURN. That caused a problem when the C++ compiler has
optimized the invocation of such methods so that it has removed its epilog
and the return address was in a completely different function.

The Regressions/coreclr/GitHub_88128/test88128 was failing due to this
om macOS arm64.

I have removed all the NORETURN markings from these methods to fix it.

BrzVlad commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -4399,7 +4394,7 @@ void DECLSPEC_NORETURN DispatchExSecondPass(ExInfo *pExInfo)
     CallCatchFunclet(pExInfo->m_exception, (BYTE *)pCatchHandler, pFrameIter->m_crawl.GetRegisterSet(), pExInfo);
     // CallCatchFunclet will resume after the catch and never return here.
-    UNREACHABLE();
+    //UNREACHABLE();
```
To better understand code flow, I believe it would be better to preserve all the `UNREACHABLE` statements as a comment, also explaining that this is a workaround for some unwinding issue.

janvorli replied:
Make sense. I guess that it would be even better to define a new kind of UNREACHABLE macro that would assert if the location is reached by accident.

janvorli committed:
Modify the unwind instead and keep the unreachable stuff untouched

janvorli commented:
@BrzVlad in the end, I've decided to solve the problem by adjusting the instruction pointer before the `PAL_VirtualUnwind` call instead. It seems the native EH does it in a similar way.

janvorli committed:
Redo the change based on how LLVM libc++ handles the case

janvorli committed:
Make the behavior LLVM vs HP libunwind specific instead of Apple

janvorli committed:
Fix the check

The handling of the case when the frame is a hardware exception
frame needs to stay the same for the LLVM libunwind too.
